src/transform/src/canonicalization/flat_map_elimination.rs

```diff
@@ -42,70 +45,49 @@ impl crate::Transform for FlatMapElimination {
-                    if with_ordinality {
-                        *relation = relation.take_dangerous().map_one(MirScalarExpr::literal(
-                            Ok(Datum::Int64(1)),
-                            ScalarType::Int64,
-                        ));
-                    }
+                    return;
+                }
+            }
+            // For all other table functions, check for all arguments being literals.
+            let mut args = vec![];
+            for e in exprs {
+                match e.as_literal() {
+                    Some(Ok(datum)) => args.push(datum),
+                    // Give up if any arg is not a literal, or if it's a literal error.
+                    _ => return,
                 }
-            } else if is_supported_unnest(func) {
-                let func = func.clone();
-                let exprs = exprs.clone();
-                use mz_expr::MirScalarExpr;
-                use mz_repr::RowArena;
-                if let MirScalarExpr::Literal(Ok(row), ..) = &exprs[0] {
-                    let temp_storage = RowArena::default();
-                    if let Ok(mut iter) = func.eval(&[row.iter().next().unwrap()], &temp_storage) {
-                        match (iter.next(), iter.next()) {
-                            (None, _) => {
-                                // If there are no elements in the literal argument, no output.
-                                relation.take_safely(None);
-                            }
-                            (Some((row, Diff::ONE)), None) => {
-                                assert_eq!(func.output_type().column_types.len(), 1);
-                                *relation =
-                                    input.take_dangerous().map(vec![MirScalarExpr::Literal(
-                                        Ok(row),
-                                        func.output_type().column_types[0].clone(),
-                                    )]);
-                                if with_ordinality {
-                                    *relation =
-                                        relation.take_dangerous().map_one(MirScalarExpr::literal(
-                                            Ok(Datum::Int64(1)),
-                                            ScalarType::Int64,
-                                        ));
-                                }
-                            }
-                            _ => {}
-                        }
-                    };
+            }
+            let temp_storage = RowArena::new();
+            let (first, second) = match func.eval(&args, &temp_storage) {
+                Ok(mut r) => (r.next(), r.next()),
+                // don't play with errors
+                Err(_) => return,
+            };
+            match (first, second) {
+                // The table function evaluated to an empty collection.
+                (None, None) => {
```

Comment:
Small nit, but could this be `(None, _)`? Idk that we actively fuse our iterators, and I think the only thing that matters is that the first call to `next()` returns a `None`. Lightly weird to call `next()` a second time after getting a `None` back, which I think we do to fit into `match`, though .. I can't tell if that is helpful or not. Not a big deal either way; just something that struck me while trying to read this.

Reply:
It could be `(None, _)`, yes.

>  Lightly weird to call next() a second time after getting a None back, which I think we do to fit into match, though .. I can't tell if that is helpful or not.

Slightly helpful. The table func iterator holds a lifetime that originates from the current MIR node so by calling `r.next()` above makes sure `r` is dropped by the time we reach this point and we want to replace the current MIR node. It's possible to hold on to it but then you have to insert manual drop statements to make Rust happy

Reply:
Yes, my original code had manual `drop`s, but this version looks cleaner.

I've modified it to have `(None, _)`.
(I could also add a `fuse` call if we are worried about some iterator panicking on the second call. But this is hopefully unnecessary.)